File: leetcode/15.java
//Runtime: 33 ms, faster than 57.58% of Java online submissions for 3Sum.
//        Memory Usage: 47.9 MB, less than 83.74% of Java online submissions for 3Sum.
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        int cur, low, high, tmp;

        Arrays.sort(nums);
        int lastIndex = nums.length-1;
        for(int i=0; i< nums.length - 2; i++) {

            if(i>=1 && nums[i] == nums[i-1]) continue;

            cur = -nums[i];
            low = i+1;
            high = lastIndex;

            while(low < high) {
                tmp = nums[low] + nums[high];
                if(tmp == cur) {
                    res.add(Arrays.asList(nums[i], nums[low], nums[high]));
                    low++;
                    high--;
// OK also
//                     while(low<lastIndex && nums[low] == nums[low - 1])
//                         low++;
//                     while(high>0 && nums[high] == nums[high + 1])
//                         high--;

                    while(low<high && nums[low]==nums[low-1])
                        low++;

                    while(low<high  && nums[high]== nums[high + 1])
                        high--;
                }
                else if(tmp > cur) {
                    high--;
                }
                else{
                    low++;
                }
            }
        }
        return res;

    }
}

File: leetcode/15.cpp
//Runtime: 84 ms, faster than 99.69% of C++ online submissions for 3Sum.
//Memory Usage: 14.6 MB, less than 98.82% of C++ online submissions for 3Sum.
class Solution {
public:
	vector<vector<int>> threeSum(vector<int>& num) {
		vector<vector<int> > res;
		std::sort(num.begin(), num.end());
		for (int i = 0; i < num.size(); i++) {
			int target = -num[i];
			int front = i + 1;
			int back = num.size() - 1;
			while (front < back) {
				int sum = num[front] + num[back];
				if (sum < target)
					front++;
				else if (sum > target)
					back--;
				else {
					vector<int> triplet(3, 0);
					triplet[0] = num[i];
					triplet[1] = num[front];
					triplet[2] = num[back];
					res.push_back(move(triplet));

					front++;
					while (front < back && num[front] == num[front - 1]) front++;

					back--;
					while (front < back && num[back] == num[back + 1]) back--;
				}
			}

			while (i + 1 < num.size() && num[i + 1] == num[i])
				i++;

		}
		return res;
	}
};